scripts/preprocess-docs.js

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env node
+
+/**
+ * Pre-processes Mocha's documentation
+ * 1. Update TOC
+ * 2. Dump `--help` output into "Command-Line Usage" section
+ */
+
+/**
+ * The CLI included with markdown-toc doesn't support `bullets`
+ * and `maxdepth` options, so that's why this exists.
+ */
+
+'use strict';
+
+const toc = require('markdown-toc');
+const miss = require('mississippi');
+const {exec} = require('child_process');
+const fs = require('fs');
+const {stripIndents} = require('common-tags');
+const path = require('path');
+
+const docsFilepath = path.join(__dirname, '..', 'docs', 'index.md');
+
+/**
+ * Inserts/updates a table of contents into the documentation
+ * @returns Promise<void>
+ */
+function updateTOC() {
+  return new Promise((resolve, reject) => {
+    console.error(`Updating TOC in ${docsFilepath}...`);
+
+    miss.pipe(
+      fs.createReadStream(docsFilepath, 'utf8'),
+      miss.concat(input =>
+        fs.writeFileSync(
+          docsFilepath,
+          toc
+            .insert(String(input), {
+              bullets: '-',
+              maxdepth: 2
+            })
+            .replace(/\n\n$/, '\n')
+        )
+      ),
+      err => {
+        if (err) {
+          return reject(err);
+        }
+        console.error(`Updated TOC in ${docsFilepath}.`);
+        resolve();
+      }
+    );
+  });
+}
+
+/**
+ * Inserts/updates command-line usage info into the documentation
+ * @returns Promise<void>
+ */
+function updateUsage() {
+  return new Promise((resolve, reject) => {
+    console.error(`Updating command-line usage in ${docsFilepath}...`);
+
+    const FENCE_HEAD = '```plain';
+    const FENCE_FOOT = '```';
+    const TOP_COMMENT = '<!-- usage -->';
+    const BOTTOM_COMMENT = '<!-- usagestop -->';
+    exec(
+      `${process.execPath} ${path.join(
+        __dirname,
+        '..',
+        'bin',
+        'mocha'
+      )} '--help'`,
+      (err, stdout) => {
+        if (err) {
+          return reject(err);
+        }
+        stdout = stdout.trim();
+
+        miss.pipe(
+          fs.createReadStream(docsFilepath, 'utf8'),
+          miss.concat(input => {
+            let trailingNewlines = '';
+            var m = /\n+$/.exec(input);
+            if (m) {
+              trailingNewlines = m[0];
+            }
+            const sections = input
+              .split(/(?:<!--\s+usage(?:\s*stop)?\s+-->)/g)
+              .map(v => v.trim());
+            if (sections.length > 3) {
+              throw new Error('invalid; multiple "usage" sections found');
+            }
+            if (sections.length === 3) {
```

Comment:
Be nice to have comment as to why invoking `--help` might have a different number of sections.

Reply:
that is a good point, because I had to figure it out from the code in `markdown-toc` that I ripped off to do this.

Reply:
I added some comments here